libavfilter/vf_livepeer_dnn.c

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (c) 2021 Livepeer Inc.
+ *
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+/**
+ * @file
+ * Filter implementing livepeer scene classification.
+ */
+
+#include "avfilter.h"
+#include "formats.h"
+#include "internal.h"
+#include "libavutil/opt.h"
+#include "libavutil/pixdesc.h"
+#include "libavutil/imgutils.h"
+#include "libavformat/avio.h"
+#include "libswscale/swscale.h"
+#include "dnn_filter_common.h"
+
+typedef struct LivepeerContext {
+    const AVClass *class;
+
+    DnnContext          dnnctx;         ///< DNN model, backend, I/O layer names
+    struct SwsContext   *sws_rgb_scale; ///< Used for scaling image to DNN input size and pixel format (RGB24)
+    struct AVFrame      *swscaleframe;  ///< Scaled image
+    FILE                *logfile;       ///< (Optional) Log classification probabilities in this file
+
+    int                 sample_rate;    ///< Run classification after every <sample_rate> frames
```

Comment:
In LPMS I decided to use the `select` filter for this, as that prevents doing hwdownload of frames that we won't run detection on: https://github.com/livepeer/lpms/pull/246/files#diff-4ab68a4de67768f1acb5be217d4ec78bab27b99381025beedf82c11a5fa9c7aeR313

This wasn't a problem in our old filter because we manually copied frames from HW->SW as opposed to using `hwdownload`

Reply:
Neat use of the select filter!

IIUC the typical structure of a scene classification enabled DNN filtergraph will be select -> hwdownload -> livepeer_dnn such that only the frames outputted by the select filter are downloaded from GPU to CPU and then once those frames are in CPU memory they're passed to livepeer_dnn.

If this is the case, then can we remove all sample rate and frame counting logic in this filter and just assume that the user of this filter wants the model to use all frames passed in and a user can select the frames of interest prior to applying this filter?

Reply:
That sounds like a good idea - will remove!